test/MediatR.Tests/SendTests.cs

```diff
@@ -183,4 +213,13 @@ public async Task Should_resolve_generic_void_handler()
 
         _dependency.Called.ShouldBeTrue();
     }
+
+    [Fact]
+    public async Task Should_resolve_multiple_type_parameter_generic_handler()
+    {
+        var request = new MultipleGenericTypeParameterRequest<TestClass1, TestClass2, TestClass3>();
+        await _mediator.Send(request);
+
+        _dependency.Called.ShouldBeTrue();
+    }
```

Comment:
Can we also have a test on `Should_resolve_closed_handler_if_defined`

I.e. with something like `TestClass1PingHandler : IRequestHandler<VoidGenericPing<TestClass1>>`

So for a given generic type `T` if there is no closed handler but there is an open one it will resolve that.  But if there is an open one and a closed one it should resolve the closed one.

This is how my implementation works. This allows you to have a an OpenGeneric handler which works for most use cases but then override this with specific implementations if needed.

Reply:
Yes I can make a test for that.. No problem.

Reply:
But that test might be redundant since the new mediatR registration removes the need to manually register them. And the manual registration becomes a duplicate service on the container.

Reply:
I'll have to add some type evaluator that tells mediator to ignore the test class used in a manual registration. Then I can test if that registration is there.

Reply:
Actually what will determine if a manually registered closed handler is called over a generic one will be the order in which the registrations happen.

This is because the closed `IRequest<VoidGenericPing<Pong>` service is already registered when calling IMediatR.  And if you wish to override this service registration then you will need to add your manual registration after the call to `AddMediatR`.

I added this test.. I modified the `Dependency` class to add a property called `CalledSpecific`. The only handler that modifies the value is the newly added closed concrete implementation that you defined above: 
`TestClass1PingHandler : IRequestHandler<VoidGenericPing<Pong>>`

```csharp
 [Fact]
 public async Task Should_resolve_closed_handler_if_defined()
 {
     var dependency = new Dependency();
     var services = new ServiceCollection();
     services.AddSingleton(dependency);       
     services.AddMediatR(cfg => cfg.RegisterServicesFromAssemblies(Assembly.GetExecutingAssembly()));
     services.AddTransient(typeof(IRequestHandler<VoidGenericPing<Pong>>), typeof(TestClass1PingRequestHandler));
     var serviceProvider = services.BuildServiceProvider();
     var mediator = serviceProvider.GetService<IMediator>()!;

     var request = new VoidGenericPing<Pong>();
     await mediator.Send(request);

     dependency.CalledSpecific.ShouldBeTrue();
 }

```

As you can see the manual registration of the handler follows the call to `AddMediatR`. If the manual registration is moved above the call to `AddMediatR` then the test will fail because Microsoft DI uses the last registered implementation for a given service.  

Does this meet requirements?

Reply:
Furthermore I added this test that ensures the opposite (that the open request handler is resolved when a manual registration for request/handler is not defined but a manual registration exists)

```csharp
[Fact]
public async Task Should_resolve_open_handler_if_not_defined()
{
    var dependency = new Dependency();
    var services = new ServiceCollection();
    services.AddSingleton(dependency);
    services.AddMediatR(cfg => cfg.RegisterServicesFromAssemblies(Assembly.GetExecutingAssembly()));
    services.AddTransient<IRequestHandler<VoidGenericPing<PongExtension>>, TestClass1PingRequestHandler>();
    var serviceProvider = services.BuildServiceProvider();
    var mediator = serviceProvider.GetService<IMediator>()!;

    var request = new VoidGenericPing<Pong>();
    await mediator.Send(request);

    dependency.Called.ShouldBeTrue();
    dependency.CalledSpecific.ShouldBeFalse();
}
```

Reply:
No that wouldn't meet my requirements because I would now have to manually register all closed handlers where this is handled automatically currently.

My implementation uses a function of my DI Container that allows registering types at runtime where they are missing (using this: https://jasperfx.github.io/lamar/documentation/ioc/registration/policies/)

So if a type is closed this is picked up by AddMediatR and takes precedence.  If I fire off a generic request without a closed handler the policy fires and resolves to the matching GenericHandler and closes this types.  

This adds a tiny overhead at runtime but only the first time this particular handler is needed and saves registering all possible closures (even if they are never called - which is what I believe this is doing).

Understanding more now how this works I think we need a way to opt out of this functionality as it is going to be hard to make this compatible with my solution, and I'm not sure this will work as well.

Reply:
yeah it's not hard to opt out. you can simply use the type evaluator. 

I'm pretty sure @jbogard doesn't want to use any third party DI containers such as lamar.

Reply:
I'm not really understanding your specific needs.. All I know is that this will resolve the issues of breaking changes for people who have manually registered generic handlers with more than one type parameter.

Reply:
Another possible way is to create a boolean configuration property: 

```csharp
public bool RegisterGenericHandlers { get; set; } = true;
```
then you could turn this off when calling `AddMediatR`:

```csharp
services.AddMediatR(cfg => 
{
    cfg.RegisterGenericHandlers = false;
    cfg.RegisterHandlersFromAssembliesContaining<Ping>();
});
```

Reply:
> Understanding more now how this works I think we need a way to opt out of this functionality as it is going to be hard to make this compatible with my solution, and I'm not sure this will work as well.

If anything this will just remove your dependency on lamar IoC.

Reply:
I'm not suggesting using Lamar and I'm using Lamar for other reasons.

> this will resolve the issues of breaking changes for people who have manually registered generic handlers with more than one type parameter

I'm not manually registering generic handlers though. that's my point.

I think adding this configuration now would be a good idea
```
public bool RegisterGenericHandlers { get; set; } = true;
```

Reply:
> I think adding this configuration now would be a good idea
> 
> ```
> public bool RegisterGenericHandlers { get; set; } = true;
> ```

Is there any chance that existing applications will still break when `RegisterGenericHandlers` is set to `true`? If that is the case, I would say that this should be set to `false` by default. That way existing applications would function exactly the same way they've functioned this far and if someone wants to turn on the generics support, it would be a conscious choice. Of course this is a bit late now as `12.3.0` has already been released anyway.

Otherwise the major version of the package should be updated to 13 if the feature is on by default and it might cause issues with existing implementations.

Reply:
> I'm not suggesting using Lamar and I'm using Lamar for other reasons.
> 
> > this will resolve the issues of breaking changes for people who have manually registered generic handlers with more than one type parameter
> 
> I'm not manually registering generic handlers though. that's my point.
> 
> I think adding this configuration now would be a good idea
> 
> ```
> public bool RegisterGenericHandlers { get; set; } = true;
> ```
you are most likely using lamar to register your generic handlers by scanning at runtime. That's just a guess. I don't know for sure.. but in any case:

This solution fixes the issues.. 

- The breaking change was caused because people had generic handlers with more than one type parameter defined in the assembly.
- MediatR failed to register them because it was hardcoded to only register generics with 1 type parameter.
- The PR now supports registration of generic handlers with multiple type parameters.
- The PR resolves the issues. Does it not?

Reply:
> Understanding more now how this works I think we need a way to opt out of this functionality as it is going to be hard to make this compatible with my solution, and I'm not sure this will work as well.

I think you don't really understand what this implementation is doing or how it works. I think you are scared that your app will break and you are kind of panicked.  

Rest assured.. This functionality will not require you to manually register all of your closing types.  That is what this implementation is doing.  I've posted several examples of this. I cannot be any more clear.

Reply:
> > I think adding this configuration now would be a good idea
> > ```
> > public bool RegisterGenericHandlers { get; set; } = true;
> > ```
> 
> Is there any chance that existing applications will still break when `RegisterGenericHandlers` is set to `true`? If that is the case, I would say that this should be set to `false` by default. That way existing applications would function exactly the same way they've functioned this far and if someone wants to turn on the generics support, it would be a conscious choice. Of course this is a bit late now as `12.3.0` has already been released anyway.
> 
> Otherwise the major version of the package should be updated to 13 if the feature is on by default and it might cause issues with existing implementations.

Also I prove that this won't break existing apps using mediator with the above tests:

```csharp
 [Fact]
 public async Task Should_resolve_closed_handler_if_defined()
 {
     var dependency = new Dependency();
     var services = new ServiceCollection();
     services.AddSingleton(dependency);       
     services.AddMediatR(cfg => cfg.RegisterServicesFromAssemblies(Assembly.GetExecutingAssembly()));
     services.AddTransient(typeof(IRequestHandler<VoidGenericPing<Pong>>), typeof(TestClass1PingRequestHandler));
     var serviceProvider = services.BuildServiceProvider();
     var mediator = serviceProvider.GetService<IMediator>()!;

     var request = new VoidGenericPing<Pong>();
     await mediator.Send(request);

     dependency.CalledSpecific.ShouldBeTrue();
 }
 ```

 1. If end users are not using generic handlers then they have nothing to worry about.
 2. if users did have generic handlers then they most likely had to manually register them or use a third party DI container like lamar, or autofac that will attempt to locate the closing types at runtime.  They too will not have any issues.
 
 The above test proves 2 to be accurate since mediatR registers all handlers correctly and then I manually register the handler.  Everything is working as expected and there are no errors.
 
 I'm confident that there doesn't need to be an opt in strategy and that an opt out strategy will suffice. I'm also confident that there doesn't need to be a major version increase.

Reply:
I've added an opt out functionality.. marking this as resolved.